plutus-chain-index-core/test/MarconiSpec.hs

```diff
@@ -0,0 +1,320 @@
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TupleSections  #-}
+
+module MarconiSpec where
+
+import Control.Monad (replicateM)
+import Data.Coerce (coerce)
+import Data.Int (Int64)
+import Data.Map (Map)
+import Data.Ratio (Ratio, (%))
+import Data.Set qualified as S
+import Numeric.Natural (Natural)
+
+import Hedgehog
+import Hedgehog.Gen qualified as Gen
+import Hedgehog.Range qualified as Range
+import Test.Tasty (TestTree, testGroup)
+import Test.Tasty.Hedgehog (testProperty)
+
+import Cardano.Api
+import Cardano.Api qualified as Api
+import Cardano.Api.Shelley
+import Gen.Cardano.Api.Metadata qualified as CGen
+import Gen.Cardano.Api.Typed qualified as CGen
+import Plutus.V1.Ledger.Api qualified as Plutus
+
+import Marconi.Index.ScriptTx qualified as ScriptTx
+
+tests :: TestTree
+tests = testGroup "Marconi"
+  [ testProperty "prop_script_hashes_in_tx_match" getTxBodyScriptsRoundtrip ]
+
+getTxBodyScriptsRoundtrip :: Property
+getTxBodyScriptsRoundtrip = property $ do
+  nScripts <- forAll $ Gen.integral (Range.linear 5 500)
+  AnyCardanoEra (era :: CardanoEra era) <- forAll $
+    Gen.enum (AnyCardanoEra ShelleyEra) maxBound
+
+  txIns <- replicateM nScripts $ forAll CGen.genTxIn
+  witnessesHashes <- replicateM nScripts $ forAll $ genWitnessAndHashInEra era
+
+  let (witnesses, scriptHashes) = unzip witnessesHashes
+      collateral = case collateralSupportedInEra era of
+        Just yes -> TxInsCollateral yes txIns
+        _        -> TxInsCollateralNone
+
+  txBody <- forAll $ genTxBodyWithTxIns era (zip txIns $ map BuildTxWith witnesses) collateral
+  let hashesFound = map coerce $ ScriptTx.getTxBodyScripts txBody :: [ScriptHash]
+  assert $ S.fromList scriptHashes == S.fromList hashesFound
+
+genTxBodyWithTxIns
```

Comment:
Doesn't the node export generators for these data structures already?

Reply:
Not exactly. They are very limited and would really need to be reworked on. Again, it would make sense to put these generators in a common package so that they can be pushed upstream.

Reply:
My concern is that generators are pretty difficult to get right. You need to think about efficient ways of shrinking them, and what statistical distribution you need to test your scenarios, and how to check that the distribution you selected is actually testing those scenarios. I am not sure we are the ones that should be doing this.

Reply:
I agree. The problem is that this is *not* a priority for the node team. We can probably do what the Hydra team did, use the generators from `cardano-ledger` and reconvert the datatype to a cardano-api type.

Ultimately, cardano-api would need to have good generators in them, and I think this is a cross-team effort (Hydra, Djed, Plutus Tools). So I think it's okay for us to start contributing in that direction IMO.